### ISSUE TITLE
Addition of "extends" top level property to enable simple configuration inheritance

### ISSUE THREAD

**Issue by greggroth** [microsoft/dev-container-spec#22]
# Problem

Multiple teams collaborating on a common codebase may have different dependency or setup needs and currently this is done by sharing a single `devcontainer.json` configuration with all of their individual needs combined. https://github.com/microsoft/dev-container-spec/issues/6 describes support for multiple configuration files, but there isn't set a way to consolidate the shared configuration into a single file.

# Proposed Solution

Introduce a new top level key `"extends"` with a value that is a relative a file path within the same repository to a "parent" devcontainer configuration. The configurations will be merged using the [same rules applied by a Docker Compose overrides file](https://docs.docker.com/compose/extends/#adding-and-overriding-configuration).

```json5
// .devcontainer/defaults.json
{
    "name": "example/project",
    "forwardPorts": [80, 5432],
    "hostRequirements": {
        "storage": "64gb",
        "memory": "16gb"
    }
}

// .devcontainer/project.json
{
    "extends": "defaults.json",
    "forwardPorts": [2222],
    "hostRequirements": {
        "memory": "32gb"
    },
   "onCreateCommand": ".devcontainer/on-create-command.sh",
}

// Results in
{
    "name": "example/project",
    "forwardPorts": [80, 5432, 2222],    // <-- Array values are the UNION
    "hostRequirements": {
        "storage": "64gb",
        "memory": "32gb"                          // <-- Basic types overwrite
    },
   "onCreateCommand": ".devcontainer/on-create-command.sh",   // <-- New keys are added
}
```

# Future 

## Add support for referring to configuration outside the repository

Users may have a use-case for keeping some shared configuration in a separate repository. The described solution does not support this, but support for this could be added.

## Add support for JSON Schema/Open API 3.0 style "$ref" document imports

The described solution is opinionated about how to merge a document, which may not be a good fit for everyone's needs (see https://github.com/docker/compose/issues/3729). A more sophisticated document reference method could be introduced to give users more control to import documents within objects or arrays and to reference objects within imported documents. For example: 

Given a file "defaults.json":

```json5
// .devcontainer/defaults.json
{
    "name": "microsoft/foo",
    "extensions": [
        "/root/hello-githug.vsix"
    ],
    "forwardPorts": [80, 5432]
    "hostRequirements": {
        "storage": "64gb",
        "memory": "32gb"
    },
    "portsAttributes": {
        "80": {
            "label": "web"
        },
        "5432": {
            "label": "postgres"
        }
    }
}
```

If a team would like to introduce a team-specific configuration that adds ports for Redis, they can add a new configuration:

```json5
// .devcontainer/redis-team.json
{
    "$ref": "defaults", 
    "extensions": [],
    "forwardPorts": [ { "$ref": "defaults#/forwardPorts" }, 6379], // Explicit merge behavior
    "hostRequirements": {
        "memory": "64gb"    // Require more memory
    },
    "portAttributes": {
        "$ref": "defaults#/portAttributes",                        // Explicit merge behavior
        "6379": {
            "label": "redis"
        }
    }
}
```

Resulting in a final configuration:

```json5
{
    "name": "microsoft/foo",
    "extensions": [],				
    "forwardPorts": [80, 5432, 6379],
    "hostRequirements": {
        "storage": "64gb",
        "memory": "64gb"
    },
    "portAttributes": {
        "80": {
            "label": "web"
        },
        "5432": {
            "label": "postgres"
        }
        "6379": {
            "label": "redis"
        }
    }
}
```

**Comment by Chuxel** [comment]
> Users may have a use-case for keeping some shared configuration in a separate repository. The described solution does not support this, but support for this could be added.

Yeah I think this is a worthy thing to consider given https://github.com/microsoft/vscode-remote-release/issues/3279.  That said, it might require the result of #7 first since this could affect how you'd make these kinds of references.  Agree that a first implementation could be local references to start, however.

In terms of the specifics though, the example above shows `.devcontainer/project.json` which wouldn't be auto-detected currently.  I assume that would be either `.devcontainer/devcontainer.json` or follow the folder structure from #6?

**Comment by greggroth** [comment]
> In terms of the example though, it shows .devcontainer/project.json which wouldn't be auto-detected currently. I assume that would be either .devcontainer/devcontainer.json or follow the folder structure from https://github.com/microsoft/dev-container-spec/issues/6?

That's right -- this is an oversight in the sample and wasn't mean to deviate from expecting that the file is named `devcontainer.json`. I'll update it.

**Comment by joshspicer** [comment]
toying around with a simple example repo: https://github.com/joshspicer/extends

**Comment by phorcys420** [comment]
sorry to be that guy but, is there any news on this topic ?

**Comment by acdha** [comment]
The main use I have for this is similar to #305: I have a project using a dev container which has a base image from our common registry. If I want to override that base image currently (for example to use an ARM image instead of x86-64), I have to edit `devcontainer.json` and remember not to commit the changes. It would be useful if I could have an override file which would change a Docker build argument but otherwise keep behaviour inline with everything else.

**Comment by cpqoc** [comment]
I have a situation where I'm dealing with 200+ repos and I am ideally looking to have this be a way to standardize and roll out updates to our teams development environment while being able to make repo specific overrides depending on project (software version, any strange config, etc).

It's working perfectly using the [alternative repo configuration folder](https://code.visualstudio.com/docs/devcontainers/create-dev-container#_alternative-repository-configuration-folders) besides the fact that I can't abstract out any of the shared config so there's a ton of duplicated config and lot of room for things to drift project to project

We're trying to centralize our tooling and local build so onboarding both new staff and new projects is as easy as install Code, open the repo, and build the container when prompted.

Overrides would save me from duplicating config in hundreds of files

I'm hoping I'm missing something that already exists, but overrides seem to be the missing piece I've been looking for in the docs

**Comment by holchan** [comment]
I have same problem, multiple repos each one with their own isolated dev container. to be worked as standalone and a main repo/orchestrator for them all... The goal is to have only one extended docker-compose.yml and a extended devcontainer.json since the only thing i need to change on them is the:

  "dockerComposeFile": "path/to/docker-compose.yml",

Right now i have to recreate multiple exact devcontainer.json to change this one line.

**Comment by holchan** [comment]
For my case i found the solution.

https://containers.dev/implementors/json_reference/

You can use LABEL dev.containers.metadata into the Dockerfile itself.

**Comment by Roboticsense00** [comment]
Simple config would be knowelege of the labeled files and what lays within each one, 
Backup!
Sort!

**Comment by neilime** [comment]
Hi, as a maintainer of many (many) projects, I tend to maintains dozens of devcontainer configs which are differing just a little from each other. 

The "extends" property will be perfect to have only few devcontainer base and extending for every projects.

Any news on this implementation?

**Comment by karpulix** [comment]
This is how I solved this problem for myself, but an `extends` feature would still be useful.

The root `devcontainer-common.json` is a template with variable`${localWorkspaceFolderBasename}`. For any new project, create a symlink to this file in its `.devcontainer` folder as `devcontainer.json`.

All projects share the root `docker-compose.yml` and `Dockerfile`.

### File Structure

```
/
├── devcontainer-common.json  # Template
├── docker-compose.yml        # Shared
├── Dockerfile                # Shared
│
├── project-a/
│   └── .devcontainer/
│       └── devcontainer.json   # -> ../../devcontainer-common.json
│
└── project-b/
    └── .devcontainer/
        └── devcontainer.json   # -> ../../devcontainer-common.json
```

1.  Create `project-b/.devcontainer`.
2.  Create a symlink: `ln -s ../../devcontainer-common.json project-b/.devcontainer/devcontainer.json`.

## Example `devcontainer-common.json`

```json
{
    "dockerComposeFile": [
        "../../docker-compose.yml"
    ],
    "extensions": [
        "ms-azuretools.vscode-docker"
    ],
    "features": {
        "ghcr.io/devcontainers/features/node:1": {
            "nvm": true,
            "version": "lts"
        }
    },
    "name": "🐳 ${localWorkspaceFolderBasename}",
    "remoteUser": "vscode",
    "service": "devcontainer",
    "settings": {
        "terminal.integrated.shell.linux": "/bin/bash"
    },
    "workspaceFolder": "/x/${localWorkspaceFolderBasename}"
}
```

**Comment by fsw0422** [comment]
There seems to be [a PR](https://github.com/devcontainers/cli/pull/311) that has been opened since 2 years related to this.

Any chance we can reiterate on this change?

**Comment by medley56** [comment]
I've had to create an entire config merge pipeline for this purpose. I keep a collection of devcontainer configs in a separate repo. Most of them use basically the same devcontainer.json template but add little tweaks for each project. I've modeled devcontainer.json in Pydantic and I actually merge together configs to generate final devcontainer.json files from a common template. 

It would be great to have this supported in the spec so I can get rid of all my jank.

**Comment by Falco20019** [comment]
Same problem here. I want to have a base spec that is also used by the CI and an extended one with some additional `postCreateCommand` and `postStartCommand` that take some time and only make sense for the devs but would significantly slow-down the CI.

**Comment by thebitbesi** [comment]
I have a huge list of "customizations" (extensions and settings for vscode) that I share among multiple containers, and it would be nice to have a way to share it across them (they already use a common docker-compose file)

**Comment by SamuelFrost** [comment]
Update for subscribers: implementation/docs are in open PRs — [CLI #1306](https://github.com/devcontainers/cli/pull/1306) and [spec #772](https://github.com/devcontainers/spec/pull/772). Feedback welcome on either; aiming to align behavior and spec text before merge.

**Comment by aioue** [comment]
> Update for subscribers: I'm working on my own open source project that utilizes devcontainers ([super_projects](https://github.com/SamuelFrost/super_projects)). So I decided to pick this up and implement it. Implementation/docs are in open PRs: CLI [devcontainers/cli#1306](https://github.com/devcontainers/cli/pull/1306) and spec [#772](https://github.com/devcontainers/spec/pull/772). Feedback with the intention to get to where we can merge this feature is welcome.

I'd aim to get the spec maintainers from this repo involved at the earliest opportunity, or you risk doing the whole [xkcd standards](https://xkcd.com/927/) dance.
